zone/spells.cpp

```diff
@@ -224,17 +224,9 @@ bool Mob::CastSpell(uint16 spell_id, uint16 target_id, CastingSlot slot,
 	if (spellbonuses.NegateIfCombat)
 		BuffFadeByEffect(SE_NegateIfCombat);
 
-	if(IsClient() && GetTarget() && IsHarmonySpell(spell_id))
-	{
-		for(int i = 0; i < EFFECT_COUNT; i++) {
-			// not important to check limit on SE_Lull as it doesnt have one and if the other components won't land, then SE_Lull wont either
-			if (spells[spell_id].effectid[i] == SE_ChangeFrenzyRad || spells[spell_id].effectid[i] == SE_Harmony) {
-				if((spells[spell_id].max[i] != 0 && GetTarget()->GetLevel() > spells[spell_id].max[i]) || GetTarget()->GetSpecialAbility(IMMUNE_PACIFY)) {
-					InterruptSpell(CANNOT_AFFECT_NPC, 0x121, spell_id);
-					return(false);
-				}
-			}
-		}
+	if (!HarmonySpellLevelCheck(spell_id, target_id)) {
```

Comment:
I would still gate this logic with `IsHarmonySpell`

CastSpell is a bit of a hot path and we are potentially unnecessarily running entity fetch logic and other calculation for things that we may not even be concerned with

I understand you're validating inside the function but validation should be done earlier in the call stack and not left to an edge function

```
if IsHarmonySpell
 -- HarmonySpellLevelCheck(spell_id, entity_list.GetMobID(target_id))
```